### ISSUE TITLE
Does not keep the changed language

### ISSUE THREAD

**Issue by matusmacko** [mcamara/laravel-localization#300]
I have installed new Laravel application (5.22) and added laravel-localization package. When I try to change the app language it changes it, however the application does not keep it set. After another request it is changed back to the language which was set before. Could it be connected with some kind of Symfony or Laravel update?  It is used to work in a proper way a few days ago.

Update - when I run Session::get('locale') it returns null. It looks like the session does not change at all.

Thank you

**Comment by MarcGaumont** [comment]
Im having the same problem... 
Its like as if the localeSessionRedirect does not work.

If im in my french language login screen and login, Im redirected to the english home page...

**Comment by MarcGaumont** [comment]
Im wondering if @arcanedev-maroc package that is based on this works.

**Comment by arcanedev-maroc** [comment]
The package : https://github.com/ARCANEDEV/Localization

**Comment by MarcGaumont** [comment]
@arcanedev-maroc Any issues with storing the session ? My redirections to the homepage when logging in from french are always back to the EN.  <https://laracasts.com/discuss/channels/laravel/localesessionredirect-mcamara>

**Comment by MarcGaumont** [comment]
@arcanedev-maroc Merci beaucoup je vais vérifier le tout ce soir en espérant ça fonctionne ;)

**Comment by SolveSoul** [comment]
Any updates on this issue?

**Comment by MarcGaumont** [comment]
I seem to have gotten this working, but using @arcanedev-maroc  package... His last comments in my Laracasts thread seem to have helped but I only had 20 min last night to check it out havent officially replied yet.

In my case I needed for sure to exclude the home '/' route from the translated routes. Also needed to extract register / login POSTs from the translated routes. Maybe theres a way around this but for now its the only way I manage to get this working.

So lets say I hover over the "Login" button, in the bottom left of my browser I see website.app/login ... but If I click it I am successfully redirected to website.app/fr/connexion ... and when I login I get redirect back to the home page in french. So I guess thats a big WIN :)

But Ill be following up posting my workings routes file for his package. Now I dont know how this translates back to the mcamara package but I will try to get that one working as well. 

The only problem I see with this is SEO wise... If the crawler in French crawls the site and is seeing duplicate links Im not sure how thats all going to workout... Ill be posting about it on laracasts

<https://laracasts.com/discuss/channels/laravel/localesessionredirect-mcamara>

**Comment by hectorsc** [comment]
Hi guys, Is this problem solved?, because I have the same problem. In Laravel 5.1 works, but not in laravel 5.2. 
Does anyone know how to fix it? :(

**Comment by elboudi** [comment]
i have the same probleme solution is change your midllware LaravelLocalizationRedirectFilter

if ( !empty( $locales[ $localeCode ] ) )
            {
                if ( $localeCode === $defaultLocale && $hideDefaultLocale )
                {
                    $redirection = app('laravellocalization')->getNonLocalizedURL();
                }
            } else if ( $currentLocale !== $defaultLocale || !$hideDefaultLocale )
            {
                // If the current url does not contain any locale
                // The system redirect the user to the very same url "localized"
                // we use the current locale to redirect him
                $redirection = app('laravellocalization')->getLocalizedURL(session('locale'),$request->path());
            }

**Comment by zachleigh** [comment]
Had this same problem. Fixed it by nesting the localization route group inside the web middleware route group. Originally, I had the localization route group as the top level group.      
This works for me:
```
Route::group(['middleware' => ['web']], function () {
    Route::group([
        'prefix' => LaravelLocalization::setLocale(),
        'middleware' => ['localeSessionRedirect', 'localizationRedirect'],
    ], function () {
        // Routes
    });
});
```

**Comment by mcamara** [comment]
Have you tried to update to the latest version of the plugin? It's working on my side...

**Comment by snyper3xs** [comment]
Not working for me. Spent a couple of hours on it. Reading all these comments and trying what I can. If I remove the redirection middlewares and just use the 'localize' middleware the selected language is not saved. Each page displays the default after setting with \LaravelLocalization::setLocale($code). I also save a cookie at this point and the cookie is accurate but LaravelLocalization is always default. When the redirection middlewares are added to the route group the site works find if the language code is in the URL. If the code is removed from the URL it is always redirected to a URL with the default code rather than the last select code. Seems it is not being stored.

**Comment by snyper3xs** [comment]
I also had a search URL which is causing issues. Is use LaravelLocalization::transRoute('routes.search') to translate the route. However it does not add any language code to the URL. The search is run using get URL parameters. So the search is run and the localizsation picks it up as missing a language code then redirects it to the default. The extra problem here is that the parameters are missing. So the language is not the last selected and the search parameters disappear. The only way I can see to fix this is to manually add in the language code the URL. However I would expect the transRoute to add in the code if one existed.

**Comment by Luddinus** [comment]
I'm having a similar problem but I don't know how to fix.

I'd like to save the language in the session, so if I'm in the url "/home/es" and I'm going to "/home", I want to redirect to "home/es" again, but it redirects always to default language ("home/en").

This is important because I have a facebook login, and facebook redirects to "/home", what can I do? I try middlewares etc but id does not work.

ping @mcamara

**Comment by dhiman252** [comment]
Any progress on this issue ?
I am using laravel v5.2.29 and the mcamara/laravel-localization v1.1.4 and not able to get what the issue is.

When a user visits the website the default lacale is `en` and after successful login I change the language via `LaravelLocalization::setLocale($user->locale);` in the controller but it still diaplays the default language not the user selected.

moreover the link https://github.com/mcamara/laravel-localization#sessions is not working. the session content does not exists on the page.

**Comment by cescgie** [comment]
@zachleigh code works for me. Thank you!

**Comment by Yeezus14** [comment]
@mcamara your answer works for me. Thanks

**Comment by aghayeff** [comment]
I added the the code at below to the top line inside function in `LaravelLocalizationRedirectFilter.php`

        session()->forget('locale');

This will fix the problem :)

**Comment by araspel** [comment]
Make sure that **"useAcceptLanguageHeader"**  parameter is **"false"** in vendor/mcamara/laravel-localization/src/config/config.php file.

**Comment by iwasherefirst2** [comment]
@araspel before [Version 1.3.12](https://github.com/mcamara/laravel-localization/releases/tag/1.3.12) enabling  "useAcceptLanguageHeader" and "hideDefaultLocaleInURL" would lead to the problem descriped by the OP. But it was fixed in Version 1.3.12 by https://github.com/mcamara/laravel-localization/pull/599
